core/math/basis.cpp

```diff
@@ -817,14 +817,13 @@ void Basis::get_axis_angle(Vector3 &r_axis, real_t &r_angle) const {
 #endif
 */
 	real_t angle, x, y, z; // variables for result
-	real_t epsilon = 0.01; // margin to allow for rounding errors
-	real_t epsilon2 = 0.1; // margin to distinguish between 0 and 180 degrees
+	real_t angle_epsilon = 0.1; // margin to distinguish between 0 and 180 degrees
 
-	if ((Math::abs(rows[1][0] - rows[0][1]) < epsilon) && (Math::abs(rows[2][0] - rows[0][2]) < epsilon) && (Math::abs(rows[2][1] - rows[1][2]) < epsilon)) {
+	if ((Math::abs(rows[1][0] - rows[0][1]) < CMP_EPSILON) && (Math::abs(rows[2][0] - rows[0][2]) < CMP_EPSILON) && (Math::abs(rows[2][1] - rows[1][2]) < CMP_EPSILON)) {
 		// singularity found
 		// first check for identity matrix which must have +1 for all terms
 		// in leading diagonal and zero in other terms
-		if ((Math::abs(rows[1][0] + rows[0][1]) < epsilon2) && (Math::abs(rows[2][0] + rows[0][2]) < epsilon2) && (Math::abs(rows[2][1] + rows[1][2]) < epsilon2) && (Math::abs(rows[0][0] + rows[1][1] + rows[2][2] - 3) < epsilon2)) {
+		if ((Math::abs(rows[1][0] + rows[0][1]) < angle_epsilon) && (Math::abs(rows[2][0] + rows[0][2]) < angle_epsilon) && (Math::abs(rows[2][1] + rows[1][2]) < angle_epsilon) && (Math::abs(rows[0][0] + rows[1][1] + rows[2][2] - 3) < angle_epsilon)) {
```

Comment:
There's addition here, not subtraction. So it seems like it's supposed to check if all of these are close enough to zero (since they are already determined to be close to each other pair-wise). Not sure what's the benefit of checking it like that over checking all of these for being close to zero separately.

Reply:
Can I consider this resolved? (I can't say more than kleonc)

Reply:
I'd either rework this or add a comment that basically says what kleonc said... otherwise the next contributor here will get confused again...

Reply:
Since this is just checking for an identity matrix, we can use `is_diagonal` to check that all the other parts are zero.

Reply:
@aaronfranke Don't you find the test problematic for the diagonal of 1?

Reply:
@fabriceci Yes, that code is also very bizzare. It... seems to be just straight up wrong, since it will be true for a basis with a diagonal of `(-5, 7, 1)`.

Reply:
@aaronfranke Since it sounds weird but we're not sure of anything, I suggest merging this PR which is a low risk one (that will be backported), and then, I can open another one that simplifies the condition.

Reply:
Sounds good. Also, for your follow-up PR, I just realized, we can use `is_equal_approx(Basis())` to check for an identity.

Reply:
@aaronfranke Nice one, I just wonder if the first condition (just before the identity one) can be replaced by `is_diagonal` and then we can skip the diagonal test and only check for the 1.